libraries/chain/controller.cpp

```diff
@@ -296,10 +296,12 @@ struct controller_impl {
     authorization( s, db ),
     protocol_features( std::move(pfs) ),
     conf( cfg ),
-    chain_id( cfg.genesis.compute_chain_id() ),
     read_mode( cfg.read_mode ),
     thread_pool( "chain", cfg.thread_pool_size )
    {
+      if (cfg_chain_id) {
```

Comment:
If I am understanding what you mean, this will work for the snapshot case, since either it will be provided for in the genesis_state provided by the version 2 snapshot, or in the global_property_object.  But how will it be set for starting without a snapshot?  Are you wanting chain_id set to genesis.compute_chain_id() if genesis is provided?

Reply:
The pattern I'm suggesting is to first construct `controller` (at this point the `chain_id` is not meaningful and should not be used) and then call one of the three `startup` overloads. After `startup` is called, `chain_id` will be set and can be safely accessed.

The `startup` overload that takes a snapshot reader will figure out the chain ID from the snapshot. The `startup` overload that takes a `genesis_state` (when initializing a fresh database as of genesis) computes the chain ID from that passed in `genesis_state`. The third `startup` overload that is meant to be called on an existing database state should be able to just reach into the `global_property_object` and grab the `chain_id` from there.